### PR TITLE
[SPARK-44128][BUILD] Upgrade netty to 4.1.93

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr aims to upgrade netty from 4.1.92 to 4.1.93.

### Why are the changes needed?
1.v4.1.92 VS v4.1.93
https://github.com/netty/netty/compare/netty-4.1.92.Final...netty-4.1.93.Final

2.The new version brings some bug fix, eg:
- Reset byte buffer in loop for AbstractDiskHttpData.setContent ([#13320](https://github.com/netty/netty/pull/13320))
- OpenSSL MAX_CERTIFICATE_LIST_BYTES option supported ([#13365](https://github.com/netty/netty/pull/13365))
- Adapt to DirectByteBuffer constructor in Java 21 ([#13366](https://github.com/netty/netty/pull/13366))
- HTTP/2 encoder: allow HEADER_TABLE_SIZE greater than Integer.MAX_VALUE ([#13368](https://github.com/netty/netty/pull/13368))
- Upgrade to latest netty-tcnative to fix memory leak ([#13375](https://github.com/netty/netty/pull/13375))
- H2/H2C server stream channels deactivated while write still in progress ([#13388](https://github.com/netty/netty/pull/13388))
- Channel#bytesBefore(un)writable off by 1 ([#13389](https://github.com/netty/netty/pull/13389))
- HTTP/2 should forward shutdown user events to active streams ([#13394](https://github.com/netty/netty/pull/13394))
- Respect the number of bytes read per datagram when using recvmmsg ([#13399](https://github.com/netty/netty/pull/13399))

3.The release notes as follows:
- https://netty.io/news/2023/05/25/4-1-93-Final.html

4.Why not upgrade to `4-1-94-Final` version?
Because the return value of the 'threadCache()' (from `PoolThreadCache` to `PoolArenasCache`) method of the netty Inner class used in the 'arrow memory netty' version '12.0.1' has changed and belongs to break change, let's wait for the upgrade of the 'arrow memory netty' before upgrading to the '4-1-94-Final' version.

The reference is as follows:
https://github.com/apache/arrow/blob/6af660f48472b8b45a5e01b7136b9b040b185eb1/java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java#L164
https://github.com/netty/netty/blob/da1a448d5bc4f36cc1744db93fcaf64e198db2bd/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L732-L736

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GA.

### CODE REVIEW

**Review by LuciferYang — APPROVED** [review]
+1, LGTM(pending test)

**Review by LuciferYang — CHANGES_REQUESTED** [review]
image

Seems there are api changes in the new netty version. Can you fix it or maybe we need wait until arrow-netty upgrade? @panbingkun

**Comment by panbingkun** [comment]
This version has been released on the official website, but it is not yet available in Maven Repo. Let's wait for a moment. 😄

**Comment by LuciferYang** [comment]
What I mean is that we may need to wait for the next arrow version to be compatible with the netty 4.1.94.Final

**Comment by panbingkun** [comment]
https://github.com/apache/arrow/blob/6af660f48472b8b45a5e01b7136b9b040b185eb1/java/memory/memory-netty/src/main/java/io/netty/buffer/PooledByteBufAllocatorL.java#L164

https://github.com/netty/netty/blob/da1a448d5bc4f36cc1744db93fcaf64e198db2bd/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L732-L736

**Comment by panbingkun** [comment]
> What I mean is that we may need to wait for the next arrow version to be compatible with the netty 4.1.94.Final

Let's upgrade to the `netty 4.1.93.Final` version first. After `arrow memory netty'` completes the same upgrade, we will consider `netty 4.1.94.Final` again.

File: pom.xml [hunk 1]
```diff
@@ -212,7 +212,7 @@
     <commons-cli.version>1.5.0</commons-cli.version>
     <bouncycastle.version>1.60</bouncycastle.version>
     <tink.version>1.9.0</tink.version>
-    <netty.version>4.1.92.Final</netty.version>
+    <netty.version>4.1.93.Final</netty.version>
```
Comment (LuciferYang) [inline pom.xml:215]: I think we should add some comments to inform other developers not to try upgrading to 4.1.94 and need to wait for arrow-memory-netty to upgrade together

**Comment by LuciferYang** [comment]
https://github.com/apache/arrow/pull/36211/files

arrow already upgrade Netty to 4.1.94.Final and this may be released in arrow 13.0,  I'm not sure if this is available because it also updates the grpc to 1.56.0, which is different from Spark Connect using a different grpc version

 
On the other hand, Netty 4.1.94 fixes a CVE https://github.com/apache/arrow/pull/36211, I want to know if this CVE will affect Spark?

File: pom.xml (hunk 1 above)
Comment (panbingkun) [reply pom.xml:219]: Done

**Review by LuciferYang — APPROVED** [review]
+1, LGTM (pending CI)

**Comment by LuciferYang** [comment]
friendly ping @srowen @dongjoon-hyun further review

I think we can upgrade to use netty 4.1.93 first, the versions greater than 4.1.94 need to wait for `arrow-memory-netty` to resolve compatibility issues and upgrade together.

**Comment by dongjoon-hyun** [comment]
I agree with you, @LuciferYang and @panbingkun , because 4.1.93.Final is the first release for Java 21 official support.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Thank you for working on this, @LuciferYang and @panbingkun !

**Comment by danepitkin** [comment]
FYI the next `arrow-memory-netty` release should be compatible with netty <4.1.94 and >=4.1.96. We plan to upgrade to 4.1.96 in `arrow-memory-netty` v13: https://github.com/apache/arrow/pull/36926

**Pull request closed without merging**